napari/_vispy/canvas.py

```diff
@@ -332,8 +332,12 @@ def _map_canvas2world(
             of the viewer.
         """
         nd = self.viewer.dims.ndisplay
-        transform = self.view.scene.transform
-        mapped_position = transform.imap(list(position))[:nd]
+        transform = self.view.camera._scene_transform
+        mapped_position = transform.imap(list(position))
+        mapped_position = (
+            mapped_position[0:nd] / mapped_position[nd]
+        )  # cartesian to homogeneous coordinates
```

Comment:
I would put comments above the code for readability (should prevent this multiline nonsense from the black formatter):

```suggestion
        # cartesian to homogeneous coordinates
        mapped_position = mapped_position[0:nd] / mapped_position[nd]
```

Same thing happens lower down!

Reply:
Now testing locally, this is wrong. It breaks in 2D, because it makes no sense in 2D (and ends up dividing by zero). This should only happen in 3D mode!

Reply:
I think in 2D, the coordinates are not represented by homogeneous, thus the last column is left 0.

Reply:
I think in 2D, the coordinates are not homogeneous, thus the last column is left 0.